backend/wmg/pipeline/summary_cubes/calculate_markers.py

```diff
@@ -0,0 +1,620 @@
+        Array of sum of squared expressions in target pop for each gene
+    n1 - np.ndarray (1 x M)
+        Number of cells in target pop for each gene
+    sum2 - np.ndarray (N x M)
+        Array of sum expression in context pops for each gene
+    sumsq2 - np.ndarray (N x M)
+        Array of sum of squared expressions in context pops for each gene
+    n2 - np.ndarray (N x M)
+        Number of cells in context pops for each gene
+
+    Returns
+    -------
+    pvals_adj - np.ndarray
+        adjusted p-values for each comparison for each gene
+    effects - np.ndarray
+        effect sizes for each comparison for each gene
+    """
+    with np.errstate(divide="ignore", invalid="ignore"):
+        mean1 = sum1 / n1
+        meansq1 = sumsq1 / n1
+
+        mean2 = sum2 / n2
+        meansq2 = sumsq2 / n2
+
+        for i in [mean1, meansq1, mean2, meansq2]:
+            i[np.isnan(i)] = 0
+
+        var1 = meansq1 - mean1**2
+        var1[var1 < 0] = 0
+        var2 = meansq2 - mean2**2
+        var2[var2 < 0] = 0
+
+        var1_n = var1 / n1
+        var2_n = var2 / n2
+        sum_var_n = var1_n + var2_n
+        dof = sum_var_n**2 / (var1_n**2 / (n1 - 1) + var2_n**2 / (n2 - 1))
+        dof[np.isnan(dof)] = 1
+        tscores = (mean1 - mean2) / np.sqrt(sum_var_n)
+        effects = (mean1 - mean2) / np.sqrt(((n1 - 1) * var1 + (n2 - 1) * var2) / (n1 + n2 - 1))
+
+    tscores[np.isnan(tscores)] = 0
+    pvals = stats.t.sf(tscores, dof)
+    pvals[effects < 0] = np.nan
+    return pvals, effects
+
+
+def _post_process_stats(genes, pvals, effects, nnz, test="ttest", min_num_expr_cells=25, percentile=0.15, n_markers=10):
+    """
+    Process and aggregate statistics into the output dictionary format.
+
+    Arguments
+    ---------
+    genes - list
+        List of genes corresponding to each p-value and effect size
+
+    pvals - np.ndarray
+        N x M array of p-values for each comparison and each gene
+
+    effects - np.ndarray
+        N x M array of effect sizes for each comparison and each gene
+
+    nnz - np.ndarray
+        1 x M array of number of nonzero expressions for target population (> 0)
+
+    test - str, optional, default "ttest"
+        The statistical test used ("ttest" or "binomtest").
+
+    min_num_expr_cells - int, optional, default 25
+        The minimum number of nonzero expressing cells required for marker genes
+
+    percentile - float, optional, default 0.15
+        The percentile of effect sizes to select as the representative effect size.
+
+    n_markers - int, optional, default 10
+        Number of top markers to return. If None, all marker genes with effect size > 0 are returned.
+    """
+    zero_out = nnz.flatten() < min_num_expr_cells
+    effects[:, zero_out] = 0
+    pvals[:, zero_out] = 1
+    # aggregate
+    effects = np.nanpercentile(effects, percentile * 100, axis=0)
+    pvals = np.array([stats.combine_pvalues(x[np.invert(np.isnan(x))] + 1e-300)[-1] for x in pvals.T])
```

Comment:
under what circumstances do we have `nan` here?

Reply:
If tscore for a particular group/gene combo is NaN. And tscore is NaN if population size for a particular group/gene combo is 0. These comparisons are excluded from aggregation.

Reply:
Got it, then should we omit the associated effect sizes as well for the percentile calculation in line 393? Are they already handled properly?

Reply:
[np.nanpercentile](https://numpy.org/doc/stable/reference/generated/numpy.nanpercentile.html) ignores NaN's, so these nan comparisons are properly excluded.